modules/dnn/src/cuda/math.cu

```diff
@@ -0,0 +1,243 @@
+        template <> __device__ float max(float x, float y) { return fmaxf(x, y); }
+        template <> __device__ double max(double x, double y) { return fmax(x, y); }
+
+        template <class T> __device__ T min(T x, T y);
+        template <> __device__ float min(float x, float y) { return fminf(x, y); }
+        template <> __device__ double min(double x, double y) { return fmin(x, y); }
+
+        template <class T> __device__ T log1p(T val);
+        template <> __device__ float log1p(float val) { return log1pf(val); }
+        template <> __device__ double log1p(double val) { return ::log1p(val); }
+
+        template <class T> __device__ T log1pexp(T val);
+        template <> __device__ double log1pexp(double val) {
+            if (val <= -37)
+                return exp(val);
+            else if (-37 < val && val <= 18)
+                return log1p(exp(val));
+            else if (18 < val && val <= 33.3)
+                return val + exp(-val);
+            else
+                return val;
+        }
+        template <> __device__ float log1pexp(float val) { return log1pexp<double>(val); }
+
+        template <class T> __device__ T tanh(T val);
+        template <> __device__ float tanh(float val) { return tanhf(val); }
+        template <> __device__ double tanh(double val) { return ::tanh(val); }
+
+        template <class T> __device__ T pow(T val, T exp);
+        template <> __device__ float pow(float val, float exp) { return powf(val, exp); }
+        template <> __device__ double pow(double val, double exp) { return ::pow(val, exp); }
+
+        template <class T>
+        __device__ T sigmoid(T val) { return T(1) / (1 + exp(-val)); }
+    }
+
+    namespace raw {
+        template <class T>
+        __global__ void abs(span<T> dest, view<T> src) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::abs;
+                dest[i] = abs(src[i]);
+            }
+        }
+
+        template <class T>
+        __global__ void tanh(span<T> dest, view<T> src) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::tanh;
+                dest[i] = tanh(src[i]);
+            }
+        }
+
+        template <class T>
+        __global__ void sigmoid(span<T> dest, view<T> src) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::sigmoid;
+                dest[i] = sigmoid(src[i]);
+            }
+        }
+
+        template <class T>
+        __global__ void bnll(span<T> dest, view<T> src) {
+            assert(src.size() >= dest.size());
+            for (auto i : grid_stride_range(dest.size())) {
+                using detail::log1pexp;
+                dest[i] = src[i] > 0 ? src[i] + log1pexp(-src[i]) : log1pexp(src[i]);
```

Comment:
[OpenCV](https://github.com/opencv/opencv/blob/master/modules/dnn/src/layers/elementwise_layers.cpp#L834): `dstptr[i] = log(1.f + exp(-abs(x)));`
[Caffe](https://github.com/BVLC/caffe/blob/master/src/caffe/layers/bnll_layer.cpp#L16-L20):  `top_data[i] = bottom_data[i] > 0 ?
        bottom_data[i] + log(1. + exp(-bottom_data[i])) :
log(1. + exp(bottom_data[i]));`

I think these are not equivalent.

Reply:
Yes, these are not the same.

Reply:
@alalek Is this a bug in OpenCV backend? 

Caffe is consistent with the formula they use in CPU and GPU code but their documentation is misleading or there is something implied which I haven't understood.

Reply:
@YashasSamaga Good catch!

Fixed here: #14899 (will be merged into master master on this/next week)